src/WebJobs.Script.WebHost/DependencyInjection/ServiceProviderExtensions.cs

```diff
@@ -0,0 +1,202 @@
+﻿// TODO: Licensing header, and split these? - these extensions are based on Orchard (https://github.com/OrchardCMS/OrchardCore)
+// With permission from the main author (https://github.com/sebastienros)
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.AspNetCore.Hosting;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Hosting;
+
+namespace Microsoft.Azure.WebJobs.Script.WebHost.DependencyInjection
+{
+    public static class ServiceProviderExtensions
+    {
+        /// <summary>
+        /// Things we don't want to copy down to child containers because...
+        /// </summary>
+        private static readonly HashSet<Type> ChildContainerIgnoredTypes = new ()
+        {
+            typeof(IStartupFilter),        // This would re-add middlewares to the host pipeline
+            typeof(IManagedHostedService), // These shouldn't be instantiated twice
+            typeof(IHostedService),        // These shouldn't be instantiated twice
+        };
+
+        /// <summary>
+        /// Creates a child container.
+        /// </summary>
+        /// <param name="serviceProvider">The service provider to create a child container for.</param>
+        /// <param name="serviceCollection">The services to clone.</param>
+        public static IServiceCollection CreateChildContainer(this IServiceProvider serviceProvider, IServiceCollection serviceCollection)
+        {
+            IServiceCollection clonedCollection = new ServiceCollection();
+            var servicesByType = serviceCollection.GroupBy(s => s.ServiceType);
+
+            foreach (var services in servicesByType)
+            {
+                if (ChildContainerIgnoredTypes.Contains(services.Key))
+                {
+                    // Do nothing - we don't want to copy these to a child scope for reasons above
+                }
+
+                // A generic type definition is rather used to create other constructed generic types.
+                else if (services.Key.IsGenericTypeDefinition)
+                {
+                    // So, we just need to pass the descriptor.
+                    foreach (var service in services)
```

Comment:
Not cloning singleton open generic service descriptors is probably not ideal.

`IOptions<T>` (`UnamedOptionsManager<T>`) comes to mind as a common open generic service where this might have undesirable side effects because `IConfigureOptions<T>` callbacks will be rerun on a new `T` even if the options have already been resolved by the parent container.

Unfortunately, I don't come with solutions because you cannot provide a single implementation instance for the `IDisposable` case, and you cannot get the concrete type in the implantation factory in the non-`IDisposable` case. I do think the comment should be updated to indicate that just passing through singleton open generic service descriptors is problematic.

Reply:
Or maybe it's a feature because you want to be able to reconfigure options in each child container? In that case, it's kind of interesting that the one scenario the child container doesn't use the parent container's singletons, it ends up helping.

Reply:
@sebastienros is this an implication you've already thought through in Orchard?